chore: move mocks under test/mocks/

fgiudici opened the pull request:
**What this PR does / why we need it**:

**Which issue this PR fixes**

Fixes #955 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

fgiudici committed:
tests: move sdkmocks package to test/mocks/sdkmocks

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici committed:
tests: use the new package location

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici committed:
tests: move metrics mocks to test/mocks/metricsmocks

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici committed:
tests: move MockRecorder to metricsmocks pkg

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

fgiudici committed:
docker: add required test/mocks to the build env

Signed-off-by: Francesco Giudici <francesco.giudici@konghq.com>

pmalek commented on Dockerfile:
```diff
@@ -22,6 +22,7 @@ COPY modules/ modules/
 COPY controller/ controller/
 COPY pkg/ pkg/
 COPY internal/ internal/
+COPY test/mocks test/mocks
```
Why do we need to include the mocks in the final Dockerfile? Non test code shouldn't rely on these 🤔

fgiudici replied:
It is related to the metricsmock.MockRecorder implementation of metrics.Recorder : MockRecorder has been moved under test/mocks. It is used in the tests but also as the default in the  konnect.NewKonnectEntityReconciler() func, as it is an implementation with methods that do nothing.
It looked more on the safe side to leave that initialization in.
Opening a comment on the related part of the code, let's move the discussion there.

pmalek reviewed (commented):
It looks like GitHub has figured it out on its own but we might as well update the .gitattributes in https://github.com/Kong/kong-operator/blob/e0a760f4182060eb7750dfc0a5fdc2c06b44b861/.gitattributes#L3.

Other than that just one comment regarding including mocks in Dockerfile.

fgiudici commented on controller/konnect/reconciler_generic.go:
```diff
@@ -101,7 +102,7 @@ func NewKonnectEntityReconciler[
 		SyncPeriod:              consts.DefaultKonnectSyncPeriod,
 		MaxConcurrentReconciles: consts.DefaultKonnectMaxConcurrentReconciles,
-		MetricRecorder:          &metrics.MockRecorder{},
+		MetricRecorder:          &metricsmocks.MockRecorder{},
```
in NewKonnectEntityReconciler() we use MockRecorder as the default initialization for the metrics.Recorder interface: MockRecorder is an implementation with methods that do nothing, so more safe than a nil interface.
This creates anyway a dependency on the metricsmocks package in main code (not test packages).
We either do not initialize MetricRecorder (leaving it to nil) or rename the implementation dropping "Mock" and move it out of test code/package.
Since NewKonnectEntityReconciler is called only once in non-test code and with the .WithMetricRecorder() option (which overwrites anyway the MetricRecorder parameter) I would go with the former proposal.
Will push an on-top commit with the proposal.

pmalek replied:
Thanks for the detailed explanation. Now I recall all of this and I agree with your view.

Looking at the tests failures it would seem that we need to add the option in each like https://github.com/Kong/kong-operator/blob/85e364f9bcd34560feb9e966f77b14cc439e2c1e/test/envtest/konnect_entities_kongcacertificate_test.go#L44-L46 to set the metrics recorder in tests and we should be good.

fgiudici committed:
Revert "docker: add required test/mocks to the build env"

This reverts commit 148d969687e1bf4f273dec12e447f353de2f93c6.

fgiudici committed:
fixup! tests: move MockRecorder to metricsmocks pkg

fgiudici committed:
fixup! tests: move sdkmocks package to test/mocks/sdkmocks

make generate artifacts follow the zz_generated.<FILENAME> template
drop legacy .gitattributes entry

fgiudici commented:
> It looks like GitHub has figured it out on its own but we might as well update the .gitattributes in
> 
> https://github.com/Kong/kong-operator/blob/e0a760f4182060eb7750dfc0a5fdc2c06b44b861/.gitattributes#L3
> 

Dropped `pkg/clientset/** linguist-generated=true` from the .gitattributes file as it looked a left over.
Changed the files generated from mockery to follow the zz_generated.<FILENAME> template already included in .gitattributes.
